### PR TITLE
feat(builtins): add vault.cat, vault.put + refactoring [NET-489 NET-491]

### PR DESCRIPTION
**Pull request by justprosh**

### CODE REVIEW

File: crates/nox-tests/tests/vault.rs [hunk 1]
```diff
@@ -217,3 +217,42 @@ async fn load_blueprint_from_vault() {
...
+        .unwrap();
+
+    let file_share = create_file_share(&mut client).await;
+    let payload = "test-test-test".to_string();
+
+    client.send_particle(
+        r#"
+        (seq
+            (seq
+                (call relay (first_service "create_vault_file") [payload] filename)
```
Comment (folex) [inline crates/nox-tests/tests/vault.rs:237]: sure you don't need a `vault put [payload]`?

**Review by folex — APPROVED** [review]

File: crates/nox-tests/tests/vault.rs (hunk 1 above)
Comment (justprosh) [reply crates/nox-tests/tests/vault.rs:237]: yeah, it's a good idea

File: crates/nox-tests/tests/vault.rs [hunk 2]
```diff
@@ -21,6 +21,7 @@ use serde_json::json;
 
 use connected_client::ConnectedClient;
 use created_swarm::make_swarms;
+use log_utils::enable_logs;
```
Comment (justprosh) [inline crates/nox-tests/tests/vault.rs:24]:
```suggestion
```

File: crates/nox-tests/tests/vault.rs [hunk 3]
```diff
@@ -217,3 +218,41 @@ async fn load_blueprint_from_vault() {
         panic!("#incorrect args: expected a single string, got {:?}", args);
     }
 }
+
+#[tokio::test]
+async fn put_cat_vault() {
+    enable_logs();
```
Comment (justprosh) [inline crates/nox-tests/tests/vault.rs:224]:
```suggestion
```

**Comment by kmd-fl** [comment]
Can you add to the description WHY do we need to add these functions? I have no idea. I also don't see it in linear.

**Review by kmd-fl — APPROVED** [review]

File: crates/nox-tests/tests/script_storage.rs [hunk 4]
```diff
@@ -602,7 +602,7 @@ async fn add_script_from_vault_wrong_vault() {
     ).await.unwrap();
 
     if let [JValue::String(error_msg)] = result.as_slice() {
-        let expected_error_prefix = "Local service error, ret_code is 1, error message is '\"Error: Incorrect vault path `/tmp/vault/another-particle-id/script";
+        let expected_error_prefix = "Local service error, ret_code is 1, error message is '\"Error reading script file `/tmp/vault/another-particle-id/script`: Incorrect vault path";
```
Comment (folex) [inline crates/nox-tests/tests/script_storage.rs:605]:
```suggestion
        let expected_error_prefix = r#"Local service error, ret_code is 1, error message is '"Error reading script file `/tmp/vault/another-particle-id/script`: Incorrect vault path"#;
```

File: crates/nox-tests/tests/vault.rs [hunk 5]
```diff
@@ -217,3 +217,40 @@ async fn load_blueprint_from_vault() {
...
+        },
+    );
+
+    use serde_json::Value::String;
+
+    let args = client.receive_args().await.unwrap();
+    if let [String(output)] = args.as_slice() {
+        assert_eq!(*output, payload);
+    } else {
+        panic!("#incorrect args: expected a single string, got {:?}", args);
```
Comment (folex) [inline crates/nox-tests/tests/vault.rs:254]: why #?
```suggestion
        panic!("incorrect args: expected a single string, got {:?}", args);
```

File: particle-execution/src/particle_vault.rs [hunk 6]
```diff
@@ -49,11 +53,67 @@ impl ParticleVault {
...
+        Ok(contents)
+    }
+
     pub fn cleanup(&self, particle_id: &str) -> Result<(), VaultError> {
         remove_dir(&self.particle_vault(particle_id)).map_err(CleanupVault)?;
 
         Ok(())
     }
+
+    fn to_virtual_path(&self, path: &Path, particle_id: &str) -> Result<PathBuf, VaultError> {
```
Comment (folex) [inline particle-execution/src/particle_vault.rs:89]: Some documentation would be nice here

File: particle-execution/src/particle_vault.rs [hunk 7]
```diff
@@ -49,11 +53,67 @@ impl ParticleVault {
...
+        let virtual_prefix = path::Path::new(VIRTUAL_PARTICLE_VAULT_PREFIX).join(particle_id);
+        let real_prefix = self.vault_dir.join(particle_id);
+        let rest = path
+            .strip_prefix(&real_prefix)
+            .map_err(|e| WrongVault(Some(e), path.to_path_buf(), real_prefix))?;
+
+        Ok(virtual_prefix.join(rest))
+    }
+
+    fn to_real_path(&self, path: &Path, particle_id: &str) -> Result<PathBuf, VaultError> {
```
Comment (folex) [inline particle-execution/src/particle_vault.rs:99]: Some documentation would be nice here

File: crates/nox-tests/tests/vault.rs (hunk 5 above)
Comment (justprosh) [reply crates/nox-tests/tests/vault.rs:254]: idk, copy-pasted

File: particle-services/src/app_service.rs [hunk 8]
```diff
@@ -15,15 +15,15 @@
  */
 
 use crate::error::ServiceError;
-use crate::{Result, VIRTUAL_PARTICLE_VAULT_PREFIX};
+use crate::Result;
 
 use fluence_app_service::{
     AppService, AppServiceConfig, MarineConfig, MarineWASIConfig, ModuleDescriptor,
 };
+use particle_execution::VIRTUAL_PARTICLE_VAULT_PREFIX;
```
Comment (folex) [inline particle-services/src/app_service.rs:23]: Is there a way to use `ParticleVault::to_virtual_path` or `to_real_path` here?

Comment (folex) [inline particle-services/src/app_service.rs:23]: I wonder if `inject_vault` should be a part of `ParticleVault`

File: particle-services/src/app_services.rs [hunk 9]
```diff
@@ -900,18 +901,13 @@ impl ParticleAppServices {
...
-            service_id.clone(),
-            self.metrics.as_ref(),
-        )
-        .inspect_err(|_| {
-            if let Some(metrics) = self.metrics.as_ref() {
-                metrics.observe_created_failed();
-            }
-        })?;
+        let service = self
+            .create_app_service(blueprint_id.clone(), service_id.clone())
```
Comment (kmd-fl) [inline particle-services/src/app_services.rs:905]: :+1:  :clap: :tada:

**Review by kmd-fl — APPROVED** [review]

**Pull request merged**